### PR TITLE
Add printflike and scanflike to all printf/scanf like functions

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Impact
Enable more compiler time check

## Testing
Pass CI

### CODE REVIEW

File: include/nuttx/usb/usbdev_trace.h [hunk 1]
```diff
@@ -419,7 +419,7 @@ typedef uint16_t usbtrace_idset_t;
 
 /* Print routine to use for usbdev_trprint() output */
 
-typedef CODE int (*trprintf_t)(const char *fmt, ...);
+typedef CODE int (*trprintf_t)(const char *fmt, ...) printflike(1, 2);
```
Comment (pkarashchenko) [inline include/nuttx/usb/usbdev_trace.h:422]: Is `printflike` applicable to typedefs?

**Review by pkarashchenko — APPROVED** [review]

**Review by hartmannathan — COMMENTED** [review]
Do all compilers support printflike and scanflike? Or would it be better to define PRINTFLIKE and SCANFLIKE in compiler.h and use those defines instead?

**Comment by xiaoxiang781216** [comment]
> Do all compilers support printflike and scanflike? Or would it be better to define PRINTFLIKE and SCANFLIKE in compiler.h and use those defines instead?

both already come from compiler.h:
https://github.com/apache/incubator-nuttx/blob/master/include/nuttx/compiler.h#L233-L237

File: include/nuttx/usb/usbdev_trace.h (hunk 1 above)
Comment (xiaoxiang781216) [reply include/nuttx/usb/usbdev_trace.h:422]: The compiler doesn't report the error, so it should be applicable.

File: include/err.h [hunk 2]
```diff
@@ -38,20 +38,20 @@
...
 /* Likewise, and then exit with STATUS.  */
 
-void err(int status, FAR const char *fmt, ...);
-void verr(int status, FAR const char *fmt, va_list ap);
-void errx(int status, FAR const char *fmt, ...);
-void verrx(int status, FAR const char *, va_list ap);
+void err(int status, FAR const char *fmt, ...) printflike(2, 3);
+void verr(int status, FAR const char *fmt, va_list ap) printflike(2, 0);
+void errx(int status, FAR const char *fmt, ...) printflike(2, 3);
+void verrx(int status, FAR const char *, va_list ap) printflike(2, 0);
```
Comment (pkarashchenko) [inline include/err.h:55]: Shouldn't this be `sysloglike`?

Comment (xiaoxiang781216) [reply include/err.h:55]: No, since they call printf internally. But _alert/_err/_warn/_info should use sysloglike instead, please take a look.

**Review by pkarashchenko — APPROVED** [review]

File: drivers/usbdev/usbdev_trprintf.c [hunk 3]
```diff
@@ -97,10 +97,10 @@ void usbtrace_trprintf(trprintf_t trprintf, uint16_t event, uint16_t value)
     case TRACE_DEVINIT:
 #ifdef CONFIG_USBDEV_TRACE_STRINGS
       trprintf("%-18s   : %-40s %04x\n", "DCD initialize",
-               TRACE_DATA(event), "Initialized", value);
+               "Initialized", value);
 #else
       trprintf("%-18s   : %04x\n", "DCD initialize",
-               TRACE_DATA(event), value);
```
Comment (ALTracer) [inline drivers/usbdev/usbdev_trprintf.c:103]: What happened here? Does TRACE_DEVINIT event not need a TRACE_DATA code be output via some %3d here, like in the other events? Or was this someone's copy-paste error that the PR unearthed? git annotate shows a 9 year old commit touching there lines.

**Comment by ALTracer** [comment]
That's very nice for people with GCC (or Clang) toolchains, who might as well be the majority of users. I think CI uses these as well.
@xiaoxiang781216 Did this change uncloak any other warnings in code which had been using the internal printf-wrappers?

File: drivers/usbdev/usbdev_trprintf.c (hunk 3 above)
Comment (pkarashchenko) [reply drivers/usbdev/usbdev_trprintf.c:103]: Seems to be some copy-paste leftover as `TRACE_DATA` is used only in `default` case if this switch

**Comment by xiaoxiang781216** [comment]
> That's very nice for people with GCC (or Clang) toolchains, who might as well be the majority of users. I think CI uses these as well. @xiaoxiang781216 Did this change uncloak any other warnings in code which had been using the internal printf-wrappers?

All it's finding is included in this PR too. More issue is found in apps side(https://github.com/apache/incubator-nuttx-apps/pull/1220).

File: drivers/usbdev/usbdev_trprintf.c (hunk 3 above)
Comment (xiaoxiang781216) [reply drivers/usbdev/usbdev_trprintf.c:103]: As @pkarashchenko said, TRACE_DATA is used for other event but not initialization.

**Review by pkarashchenko — APPROVED** [review]
LGTM!

**Pull request merged**